### PR TITLE
Adding support to AutoValue.Id.

### PR DESCRIPTION
**Pull request by rdtorres**
This new annotation should be used when an autoValue does not want to use all its fields in the equals() and hashcode() method. So, the autovalue bean can simple annotate all fields that should be used to unique identify that bean.

Let me know if you like it, so I can add more JUnit tests to cover edge cases.

### CODE REVIEW

**Comment by rdtorres** [comment]
I used another email account to submit this pull request. How can I associate this other account to my CLA account ?

File: value/src/main/java/com/google/auto/value/processor/AutoValueProcessor.java [hunk 1]
```diff
@@ -494,7 +504,7 @@ private void fixReservedIdentifiers(Map<ExecutableElement, String> methodToIdent
   }
 
   private String disambiguate(String name, Collection<String> existingNames) {
-    for (int i = 0; ; i++) {
```
Comment (rdtorres) [inline value/src/main/java/com/google/auto/value/processor/AutoValueProcessor.java]: This change has nothing to do with the commit. But I fixed it because checkstyle was complaining about it.

File: value/src/main/java/com/google/auto/value/processor/AutoValueIdsProcessor.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Annotation processor that checks that the type that {@link com.google.auto.value.AutoValue.Id}
+ * is applied to is nested inside an {@code @AutoValue} class. The actual code generation for ids
+ * is done in {@link AutoValueProcessor}.
+ *
+ * @author Rafael Torres
+ */
+@AutoService(Processor.class)
+public class AutoValueIdsProcessor extends AbstractProcessor {
```
Comment (rdtorres) [inline value/src/main/java/com/google/auto/value/processor/AutoValueIdsProcessor.java]: I was not sure if I should create another processor to validate or if I should use AutoValueBuilderProcessor. My suggestion is to rename AutoValueBuilderProcessor to some generic name and use it to validate Builder, Validate and Id.

**Comment by eamonnmcmanus** [comment]
We explicitly say in the [documentation](https://github.com/google/auto/tree/master/value#derived-fields) that we don't support fields that don't participate in equals and hashCode. If we did want to do so, this would certainly be one way of going about it. In that case I think we would also want the ability to say, independently, whether a field appears in toString.
Pinging @kevinb9n.

**Comment by rdtorres** [comment]
I've read the documentation and I know that auto-value does not support it.
In fact,  this is the main reason behind this pull request: Add to support
to it =). The toString method can also follow a similar approach where an
annotation like "Autovalue.Printable" would indicate which fields should be
printed or not. Again, these annotations are extra, non-mandatory
annotations. So, default auto-model behaviour will be exactly the same as
we have nowadays...

Let me know what do you guys think.
